python/TestHarness/schedulers/RunPBS.py

```diff
@@ -70,15 +89,46 @@ def hasTimedOutOrFailed(self, job_data):
                     job.addCaveats('TESTHARNESS EXCEPTION')
                 return True
 
+    def _optimizeSelect(self, job):
+        """ Determine the optimal select statement for PBS """
+
+        slots = int(job.getMetaData().get('QUEUEING_NCPUS', 1))
+        all_nodes = self._getPBSResources()
+        total_cores = sum([int(cores) for cores in all_nodes])
+
+        # Is this job to big for this cluster?
+        if slots > total_cores:
+            job.addCaveats('insufficient slots')
+            job.setStatus(job.skip)
+            # continue and allow the qsub script to be created
+
+        procs_per_node = int(max(groupby(sorted(all_nodes)), key=lambda(x, v):(len(list(v)),-all_nodes.index(x)))[0])
+        num_full_chunks = slots / procs_per_node
+        remaining_ranks = slots % procs_per_node
+
+        # slots > procs_per_node and is not evenly distributable (both non-zero)
+        if num_full_chunks and remaining_ranks:
+            select_ncpus = 'select=%d:ncpus=%d+%d' % (num_full_chunks, procs_per_node, remaining_ranks)
```

Comment:
This still isn't right. The remaining ranks bit should be like the case below, not just "+n". I think a better way to generate this is to generate a "left and right" part and put them together when they are both non-zero with the plus otherwise just use one or the other halves.

Reply:
I am still confused... Let me be literal for a moment:

```python
        # get the common processor per node count
        left = int(max(groupby(sorted(all_nodes)), key=lambda(x, v):(len(list(v)),-all_nodes.index(x)))[0])

        chunks = slots / left
        right = slots % left

        # left is _always_ true
        if True and right:
            select = 'ncpus=%d+%d' % (left, right)

        else:
            select = 'ncpus=%d' % (left)

        if chunks:
            select = 'select=%d:%s' % (chunks, select)

        else:
            select = 'select=1:%s' % (select)
```
What is this 'left' part/integer you are talking about? Where does it come from?

Reply:
What I meant by "left and right" was the parts before and after the plus sign. More specifically, this is the "whole" and "partial" nodes.

Reply:
Hmmm, well that is still confusing because 'left' or 'whole' is always true (Its our processor count per node).

It is 'chunks' and 'partial' that can either be true or false... And if we have a _chunk_, then the left side of ncpus will _always_ be processor count per node.

And so, that is why I am doing the following (not caring what the left side of ncpus might be):
```python

        whole = slots / procs_per_node
        partial = slots % procs_per_node

        if whole and partial:
            select = 'select=whole:ncpus=procs_per_node+partial'
        elif partial:
            select = 'select=1:ncpus=partial'
        else:
            select = 'select=whole:npucs=procs_per_node'
```

We can do it the other way, but it takes more lines of code:

```python
        whole = slots / procs_per_node
        partial = slots % procs_per_node

        if procs_per_node and partial:
            ncpus = 'ncpus=procs_per_node+partial'
        else:
            ncpus = 'ncpus=procs_per_node'

        if whole:
            select = 'select=whole:ncpus'
        else:
            select = 'select=1:ncpus'
```